### PR TITLE
rename annotation

### LINKED ISSUE #42: Rename the async always identifier
Mode would be better than frequency.
Additionally, allow two options and a default, not just one option.

### PR DESCRIPTION
**Pull request by beemarie**
# Changes

- :gift: rename the async frequency type annotation to mode
- add some validation checking to the annotation
- :broom: add testing 

/kind cleanup

Fixes #42 

**Release Note**

```release-note

```

**Docs**

```docs

```

### CODE REVIEW

**Comment by beemarie** [comment]
@duglin @julz Happy to consider other naming schemes here, but if i remember correctly, we landed on `mode` in our last chat.

**Comment by duglin** [comment]
I like "mode", but "sometimes" is a bit awkward. Perhaps something like "conditional" or "onrequest" instead?

**Comment by beemarie** [comment]
> I like "mode", but "sometimes" is a bit awkward. Perhaps something like "conditional" or "onrequest" instead?

thanks! Agree it is awkward -- let's do conditional.

File: pkg/reconciler/ingress/ingress.go [hunk 1]
```diff
@@ -31,28 +31,34 @@ type Reconciler struct {
...
-	asyncFrequencyType              = "always.async.knative.dev"
-	publicLBDomain                  = "istio-ingressgateway.istio-system.svc.cluster.local"
-	privateLBDomain                 = "cluster-local-gateway.istio-system.svc.cluster.local"
-	producerServiceName             = "producer-service"
+	asyncSuffix            = "-async"
+	newSuffix              = "-new"
+	preferHeaderField      = "Prefer"
+	preferAsyncValue       = "respond-async"
+	preferSyncValue        = "respond-sync"
+	asyncModeAnnotationKey = "async.knative.dev/mode"
```
Comment (maximilien) [inline pkg/reconciler/ingress/ingress.go:39]: since this is a key and does not change (presumably) then it could be public?

Comment (maximilien) [reply pkg/reconciler/ingress/ingress.go:39]: And a constant...

File: pkg/reconciler/ingress/ingress_test.go [hunk 2]
```diff
@@ -146,7 +158,18 @@ func TestReconcile(t *testing.T) {
...
+			Objects: []runtime.Object{
+				ingSometimesAsync,
+			},
+			WantCreates: []runtime.Object{
+				createdIng,
+				service(defaultNamespace, testingName),
+			},
+		},
+		{
+			Name: "create new ingress with async annotation and always mode",
```
Comment (maximilien) [inline pkg/reconciler/ingress/ingress_test.go:172]: Add “... mode value” to be consistent with below (or vice versa)

**Review by maximilien — APPROVED** [review]
Few minor suggestions. Best.

**Comment by maximilien** [comment]
Also you now have conflicts. Likely a rebase to get previous merged PR.

File: pkg/reconciler/ingress/ingress.go (hunk 1 above)
Comment (beemarie) [reply pkg/reconciler/ingress/ingress.go]: already a constant, but updated key to be public.

**Review by maximilien — APPROVED** [review]
/lgtm

**Pull request merged**